src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs

```diff
@@ -1079,7 +1079,7 @@ public sealed override Cci.ITypeReference GetPlatformType(Cci.PlatformType platf
                     throw ExceptionUtilities.UnexpectedValue(platformType);
 
                 default:
-                    return GetSpecialType((SpecialType)platformType, (TSyntaxNode)context.SyntaxNode, context.Diagnostics);
+                    return GetSpecialType((SpecialType)platformType, (TSyntaxNode)context.SyntaxReference?.GetSyntax(), context.Diagnostics);
```

Comment:
>context.SyntaxReference?.GetSyntax() [](http://example.com/codeflow?start=82&length=36)

I am curious how hard would it be to avoid unwrapping the reference until that is absolutely necessary (here and in many other places). I assume syntax node is used only when we report an error.

Reply:
Seems that would lead to lot of refactoring considering how many places are changed. This PR solves this for one such place - the function pointer emit error (which unwraps the reference only when really needed) - and other places can be refactored in the future.

> unnecessary allocations of `Location` instances.

This makes me wonder - won't this whole PR lead to unnecessary allocations of `SimpleSyntaxReference` instances which just wrap `SyntaxNode`s? (Because when creating `EmitContext`, it's now often needed to do `syntaxNode.GetReference()`.) Or is it common for `SyntaxReference`s to be other than the simple implementation?

Reply:
> This makes me wonder - won't this whole PR lead to unnecessary allocations of `SimpleSyntaxReference` instances which just wrap `SyntaxNode`s? (Because when creating `EmitContext`, it's now often needed to do `syntaxNode.GetReference()`.) Or is it common for `SyntaxReference`s to be other than the simple implementation?

Yes, I was concerned about this aspect too. Essentially, for majority of consumption scenarios we will start paying for creating a ```SyntaxReference``` plus an allocation, and then paying for unwrapping it. While trying to improve one thing, we might make other things worse. And net result might be negative at the end.

I think there is another option we could consider. Slightly increase the size of ```EmitContext``` and have it store either a ```SyntaxNode``` or a ```SyntaxReference```. In order to avoid changing majority of use sites, ```EmitContext.SyntaxNode``` could be converted into a property, which unwraps the ```SyntaxReference``` when ```SyntaxNode``` is not available. The size increase can be optionally mitigated by passing ```EmitContext``` everywhere by reference.